File: C Documention/Linked List/DoubleLinkedList.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
    struct Node* prev;
};

struct Node* createNode(char data){
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode -> data = data;
    newNode -> next = newNode;
    newNode -> prev = newNode;
    return newNode;
}

void InsertFirst(struct Node** head, char data){
    struct Node* newNode = createNode(data);
    newNode -> next = *head;
    newNode -> prev = NULL;
    *head = newNode;
}

void InsertNext(struct Node** head, char data){
    struct Node* ptr = *head;
    struct Node* newNode = createNode(data);
    while (ptr -> next){
        ptr = ptr -> next;
    }
    ptr -> next = newNode;
    newNode -> prev = ptr;
}

void InsertMid(struct Node** head, char data, int pos){
    struct Node* ptr = *head;
    struct Node* newNode = createNode(data);
    int pos_next = 0;
    while (ptr != NULL && pos_next + 1 != pos){
        ptr = ptr -> next;
        pos_next += 1;
    }
    if (pos_next <= 1){
        newNode -> next = ptr -> next;
        ptr -> next = newNode;
        newNode -> prev = ptr;
        } else {
            ptr -> next = newNode -> next;
            newNode -> prev = ptr;
            newNode -> next = newNode;
        }
}

void print(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d <-> ", temp->data);
        temp = temp->next;
    }
    printf("NULL\n");
}

int main(){
    struct Node* head = NULL;

    InsertFirst(&head, 10);
    InsertNext(&head, 5);
    InsertMid(&head, 20, 1);
    print(head);
}

File: Python Documention/Campus Documention/Struct_Data/Linked-List/Double-Linked_List.py
class Node:
    def __init__(self):
        self.data = input('Data : ')
        self.prev = None
        self.next = None
        
class Double_Linked_List:
    def __init__(self):
        self.head = None
        
    def addFirst(self):
        smp = Node()
        if self.head is None:
            self.head = smp
            self.head.prev = None
            
    def addNextNode(self):
        smp = Node()
        if self.head is None:
            print('Linked list is empty!')
        else:
            ptr = self.head
            while ptr.next is not None:
                ptr = ptr.next
            ptr.next = smp
            smp.prev = ptr

    def addMidNode(self, pos):
        ptr = self.head
        smp = Node()
        pos_next = 0
        if self.head is None:
            print('Linked list is empty!')
        else:
            while ptr != None and pos_next + 1 != pos :
                ptr = ptr.next
                pos_next += 1
            if pos_next <= 1:
                smp.next = ptr.next
                smp.prev = ptr
                ptr.next = smp
            else:
                if smp.next is not None:
                    smp.next.prev = ptr
                ptr.next = smp.next
                smp.prev = ptr
                smp.next = smp
                       
    def addLast(self):
        ptr = self.head
        smp = Node()
        if self.head is None:
            print('Linked list is empty!')
        else:
            while ptr.next:
                ptr = ptr.next
            if ptr.prev is not None and ptr.next is None:
                ptr.next = smp
                smp.prev = ptr
            
    def removedFirst(self):
        ptr = self.head
        if self.head is None:
            print('Linked list is empty!')
        else:
            self.head = ptr.next
            
    def removedLast(self):
        ptr = self.head
        if self.head is None:
            print('Linked list is empty!')
        else:
            while ptr.next.next:
                ptr = ptr.next
            ptr.next = None
            
    def reverseLink(self):
        ptr = self.head
        prevNode = None
        if self.head is None:
            print('Linked list is empty!')
        else:
            while ptr is not None:
                prevNode = ptr.prev
                ptr.prev = ptr.next
                ptr.next = prevNode
                ptr = ptr.prev
            self.head = prevNode.prev
            
    def display(self):
        if self.head is None:
            print('Linked list is empty!')
        else:
            ptr = self.head
            while True:
                print(ptr.data, end = ' <<>> ')
                if ptr.next is None:
                    print('None')    
                    break    
                else:
                    ptr = ptr.next
                    
dll = Double_Linked_List()
dll.addFirst()
dll.addNextNode()
dll.addNextNode()
dll.addMidNode(2)
dll.display()